src/Form/ActionForm.php

```diff
@@ -0,0 +1,188 @@
+  }
+
+  /**
+   * Build the form elements.
+   *
+   * @param array $element
+   *   The element with the attached action form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The form state.
+   * @param array $context
+   *   The context of this form.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   */
+  public function buildForm(array &$element, FormStateInterface $form_state, array $context) {
+
+    /** @var \Drupal\Core\Field\FieldItemListInterface $items */
+    $items = $context['items'];
+    if ($items->isEmpty()) {
+      return;
+    }
+
+    $field_definition = $items->getFieldDefinition();
+
+    $entity_type = $field_definition->getSettings()['target_type'];
+
+    $actionStorage = $this->entityTypeManager->getStorage('action');
+    $actions = array_filter($actionStorage->loadMultiple(), function ($action) use ($entity_type) {
+      return $action->getType() == $entity_type;
+    });
+
+    if (empty($actions)) {
+      return;
+    }
+
+    $form_state->set('entity_reference_' . $field_definition->getName(), $context['items']);
+
+    $options = [];
+    foreach ($actions as $id => $action) {
+      $options[$id] = $action->label();
+    }
+
+    $element += ['#tree' => TRUE];
+
+    $element['entity_reference_actions'] = [
+      '#type' => 'container',
+    ];
+
+    $element['entity_reference_actions']['options'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Available actions'),
+      '#options' => $options,
+    ];
+
+    $element['entity_reference_actions']['submit'] = [
+      '#type' => 'submit',
+      '#name' => $field_definition->getName() . '_button',
+      '#value' => $this->t('Apply to referenced items'),
+      '#submit' => [[$this, 'submitForm']],
+    ];
+
+  }
+
+  /**
+   * Submit function to call the action.
+   *
+   * @param array $form
+   *   The form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The form state.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+
+    $button = $form_state->getTriggeringElement();
+    $field_name = reset($button['#array_parents']);
+
+    /** @var \Drupal\Core\Entity\Display\EntityFormDisplayInterface $form_display */
+    $form_display = $form_state->getStorage()['form_display'];
+
+    /** @var \Drupal\Core\Field\WidgetInterface $widget */
+    $widget = $form_display->getRenderer($field_name);
+
+    /** @var \Drupal\Core\Field\FieldItemListInterface $items */
+    $items = $form_state->get('entity_reference_' . $field_name);
+
+    $widget->extractFormValues($items, $form, $form_state);
```

Comment:
What happens if there are ?

a) unsaved changes (entities are in the system, but form has not been saved)
b) new references (entities have not been created yet)

Reply:
Filtered out unsaved